apache/trafficcontrol
# Debug TM, TO, TO Perl, or TR CDN-in-a-Box containers

## What does this PR (Pull Request) do?

This PR lets you use CDN-in-a-Box to debug Traffic Monitor, Traffic Ops, Traffic Ops Perl, and Traffic Router and adds configuration options so that this behavior is only enabled when desired.

## Which Traffic Control components are affected by this PR?


- CDN in a Box
- Documentation
- Traffic Monitor
- Traffic Ops

## What is the best way to verify this PR?

* Build documentation to make sure there are no warnings
* Make sure documentation is accurate and easy to follow
* Ensure RPMs and CDN-in-a-Box build
* Check whether debugging works in CDN-in-a-Box for:
    - Traffic Monitor
    - Traffic Ops
    - Traffic Ops Perl
    - Traffic Router

* Check that no debugging information or behavior is included when debug flags and variables are not used.
    - When `pkg` is used instead of `pkg -d`, or CiaB's `make` is used instead of `make debug`, the `traffic_monitor` and `traffic_ops_golang` binaries should be roughly 2 MB smaller.
    - In CiaB's `variables.env`, if `DEBUGGING_TM`, `DEBUGGING_TO`, `DEBUGGING_TO_PERL`, or `DEBUGGING_TR` are not set to true, timeouts should be short (60 seconds or less) and the services should not try to use Delve, Devel::Camelcadedb, or JPDA to run.

* Jenkins should be able to build everything. That is essentially our test for the build scripts and docker environment.

## If this is a bug fix, what versions of Traffic Control are affected?

Not a bug fix.

## The following criteria are ALL met by this PR




## Additional Information



Debugging can be useful for a developer's own analysis, but it in no way replaces tests.

Comments, feedback, suggestions, and criticism are welcome.
